### PR TITLE
Added a case in analyze-schema to check for the insufficient columns …

### LINKED ISSUE #581: Update analyze-schema to report for Primary keys columns missing Partition columns
We have report the scenario explained in https://github.com/yugabyte/yb-voyager/issues/578 in analyze-schema report

### PR DESCRIPTION
**Pull request by priyanshi-yb**
…in PRIMARY KEY constraint (#578), fixed #581

Explanation - 
1. Match regex to fetch object name, all the columns info and the partition columns info.
2. fetch the Primary key column info from all the columns info, which is present at the end in `()`.
3. check one by one each partition column present in primary column list.

### CODE REVIEW

File: yb-voyager/cmd/analyzeSchema.go [hunk 1]
```diff
@@ -89,6 +89,7 @@ var (
 	amRegex              = regexp.MustCompile(`(?i)CREATE ACCESS METHOD ([a-zA-Z0-9_."]+)`)
 	idxConcRegex         = regexp.MustCompile(`(?i)REINDEX .*CONCURRENTLY ([a-zA-Z0-9_."]+)`)
 	storedRegex          = regexp.MustCompile(`(?i)([a-zA-Z0-9_]+) [a-zA-Z0-9_]+ GENERATED ALWAYS .* STORED`)
+	createTableRegex 	 = regexp.MustCompile(`(?i)CREATE TABLE (IF NOT EXISTS)?([^,]+(?:,[^,]+){0,}, PRIMARY KEY ([^,]+(?:,[^,]+){0,})) PARTITION BY RANGE([^,]+(?:,[^,]+){0,});`)
```
Comment (rahulb-yb) [inline yb-voyager/cmd/analyzeSchema.go:92]: Change the name of the variable to refer to the actual part of the regex relating to which the error may arise, for eg. `partitionRegex` (if there are no other regex related to partition, else extend it to be more specific)

File: yb-voyager/cmd/analyzeSchema.go [hunk 2]
```diff
@@ -452,6 +453,16 @@ func checkDDL(sqlInfoArr []sqlInfo, fpath string) {
 			reportCase(fpath, "LANGUAGE C not supported yet.",
 				"", "", "FUNCTION", tbl[2], sqlInfo.formattedStmt)
 			summaryMap["FUNCTION"].invalidCount++
+		} else if regMatch := createTableRegex.FindStringSubmatch(strings.Join(strings.Fields(sqlInfo.stmt), " ")); regMatch !=nil {
```
Comment (rahulb-yb) [inline yb-voyager/cmd/analyzeSchema.go:456]: This is a very roundabout way to do a string submatch, your regex handles multiple whitespaces anyways right? Please check if it works without using `Join()` and `Fields()`, and if it does not, please modify the regex to handle this.
P.S. Utilize some publicly available [regex checker](https://regex101.com), only if needed

File: yb-voyager/cmd/analyzeSchema.go [hunk 3]
```diff
@@ -452,6 +453,16 @@ func checkDDL(sqlInfoArr []sqlInfo, fpath string) {
...
+			primaryKeyColumns := strings.Trim(regMatch[3], `()`)
+			partitionColumns := strings.Trim(regMatch[4], `()`)
+			partitionColumnsList := strings.Split(partitionColumns, ",")
```
Comment (rahulb-yb) [inline yb-voyager/cmd/analyzeSchema.go:459]: `partitionColumnsList := utils.CsvStringToSlice(....)`

**Review by rahulb-yb — COMMENTED** [review]
Noted some changes to be made, please respond/resolve them

File: yb-voyager/cmd/analyzeSchema.go [hunk 4]
```diff
@@ -452,6 +453,16 @@ func checkDDL(sqlInfoArr []sqlInfo, fpath string) {
...
+			for _, each := range partitionColumnsList {
+				if !strings.Contains(primaryKeyColumns, each) {
+					reportCase(fpath, "insufficient columns in the PRIMARY KEY constraint definition in CREATE TABLE",
+				"https://github.com/yugabyte/yb-voyager/issues/578", "Add all Partition columns to Primary Key", "TABLE", regMatch[1], sqlInfo.formattedStmt)
```
[the remark this answers is not in the document]
Comment (rahulb-yb) [reply yb-voyager/cmd/analyzeSchema.go:463]: Additionally, change the name from `each` to a more read-friendly name for future developers.

File: yb-voyager/cmd/analyzeSchema.go [hunk 5]
```diff
@@ -452,6 +453,16 @@ func checkDDL(sqlInfoArr []sqlInfo, fpath string) {
 			reportCase(fpath, "LANGUAGE C not supported yet.",
 				"", "", "FUNCTION", tbl[2], sqlInfo.formattedStmt)
 			summaryMap["FUNCTION"].invalidCount++
+		} else if regMatch := createTableRegex.FindStringSubmatch(strings.Join(strings.Fields(sqlInfo.stmt), " ")); regMatch !=nil {
+			primaryKeyColumns := strings.Trim(regMatch[3], `()`)
+			partitionColumns := strings.Trim(regMatch[4], `()`)
+			partitionColumnsList := strings.Split(partitionColumns, ",")
+			for _, each := range partitionColumnsList {
+				if !strings.Contains(primaryKeyColumns, each) {
```
[the remark this answers is not in the document]
Comment (priyanshi-yb) [reply yb-voyager/cmd/analyzeSchema.go:461]: Oh yeah, thanks for this case!

File: yb-voyager/cmd/analyzeSchema.go [hunk 6]
```diff
@@ -452,6 +454,25 @@ func checkDDL(sqlInfoArr []sqlInfo, fpath string) {
 			reportCase(fpath, "LANGUAGE C not supported yet.",
 				"", "", "FUNCTION", tbl[2], sqlInfo.formattedStmt)
 			summaryMap["FUNCTION"].invalidCount++
+		} else if regMatch := partitionColumnsRegex.FindStringSubmatch(sqlInfo.stmt); regMatch !=nil {
+			allColumns := strings.Trim(regMatch[3], `()`)
+			allColumns = strings.Trim(allColumns, " ")
```
Comment (rahulb-yb) [inline yb-voyager/cmd/analyzeSchema.go:459]: Combine lines 458 and 459, the second argument passed to trim is the cut-set of characters

File: yb-voyager/cmd/analyzeSchema.go [hunk 7]
```diff
@@ -452,6 +454,25 @@ func checkDDL(sqlInfoArr []sqlInfo, fpath string) {
...
+			allColumnsList := strings.Split(allColumns, "(")
+
+			primaryKeyColumns := allColumnsList[len(allColumnsList)-1]
+			primaryKeyColumns = strings.Trim(primaryKeyColumns, `()`)
```
Comment (rahulb-yb) [inline yb-voyager/cmd/analyzeSchema.go:463]: is there a need for this line after the trim on line 458?

File: yb-voyager/cmd/analyzeSchema.go [hunk 8]
```diff
@@ -452,6 +454,25 @@ func checkDDL(sqlInfoArr []sqlInfo, fpath string) {
...
+			partitionColumns := strings.Trim(regMatch[4], `()`)
+
+			partitionColumnsList := utils.CsvStringToSlice(partitionColumns)
+			primaryKeyColumnsList := utils.CsvStringToSlice(primaryKeyColumns)
+			for _, eachPartitionColumn := range partitionColumnsList {
+				eachPartitionColumn = strings.Trim(eachPartitionColumn, " ")
```
Comment (rahulb-yb) [inline yb-voyager/cmd/analyzeSchema.go:469]: `utils.CsvStringToSlice` cleans up whitespaces while converting to slice; not sure if this line is needed

File: yb-voyager/cmd/analyzeSchema.go [hunk 9]
```diff
@@ -452,6 +454,25 @@ func checkDDL(sqlInfoArr []sqlInfo, fpath string) {
...
+				idxInPrimaryKeyColumns := sort.SearchStrings(primaryKeyColumnsList, eachPartitionColumn)
+				if idxInPrimaryKeyColumns == len(primaryKeyColumnsList) || primaryKeyColumnsList[idxInPrimaryKeyColumns] != eachPartitionColumn {
```
Comment (rahulb-yb) [inline yb-voyager/cmd/analyzeSchema.go:471]: I don't think `primaryKeyColumnsList[idxInPrimaryKeyColumns]` would return the expected value seeing that we never explicitly sorted this slice..

File: yb-voyager/cmd/analyzeSchema.go (hunk 7 above)
Comment (priyanshi-yb) [reply yb-voyager/cmd/analyzeSchema.go:463]: Yes, it is required as one `)` closing bracket will be there at the end when I splited the allColumns with respect to `(`.

File: yb-voyager/cmd/analyzeSchema.go (hunk 9 above)
Comment (priyanshi-yb) [reply yb-voyager/cmd/analyzeSchema.go:471]: Oh yes, thanks. I assumed that this sort.SearchStrings function will sort the Slice and then search.

File: yb-voyager/cmd/analyzeSchema.go (hunk 7 above)
Comment (rahulb-yb) [reply yb-voyager/cmd/analyzeSchema.go:463]: Doesn't `allColumns = strings.Trim(allColumns, "() ")` remove all trailing and leading instances of `(`, `)` and whitespaces? We seem to be picking up only the last element from `allColumnsList`, which should have all leading instances (as you mentioned in your case, `)`) removed

Comment (priyanshi-yb) [reply yb-voyager/cmd/analyzeSchema.go]: yes it should, made the changes. Please review it again. Thanks

**Review by rahulb-yb — APPROVED** [review]
LGTM, please omit excess empty new lines in the code for the sake of continuity, and then feel free to merge this PR 👍🏽

**Pull request merged**